pkg/resource/imageconfig.go

```diff
@@ -62,26 +71,78 @@ func (gic *generatorImageConfig) Create() error {
 		},
 	}
 
-	_, err := gic.client.Images().Create(ic)
+	externalHostnames, err := gic.getRouteHostnames()
+	if err != nil {
+		return err
+	}
+	ic.Status.ExternalRegistryHostnames = externalHostnames
+
+	_, err = gic.configClient.Images().Create(ic)
+	if err != nil {
+		return err
+	}
+	// Create strips status fields, so need to explicitly set status separately
+	_, err = gic.configClient.Images().UpdateStatus(ic)
 	return err
 }
 
 func (gic *generatorImageConfig) Update(o runtime.Object) (bool, error) {
 	ic := o.(*configapi.Image)
 
+	externalHostnames, err := gic.getRouteHostnames()
+	if err != nil {
+		return false, err
+	}
+
 	modified := false
+	if !reflect.DeepEqual(externalHostnames, ic.Status.ExternalRegistryHostnames) {
+		ic.Status.ExternalRegistryHostnames = externalHostnames
+		modified = true
+	}
+
 	if ic.Status.InternalRegistryHostname != gic.hostname {
 		ic.Status.InternalRegistryHostname = gic.hostname
 		modified = true
 	}
+
 	if !modified {
 		return false, nil
 	}
 
-	_, err := gic.client.Images().UpdateStatus(ic)
+	_, err = gic.configClient.Images().UpdateStatus(ic)
 	return true, err
 }
 
 func (gic *generatorImageConfig) Delete(opts *metav1.DeleteOptions) error {
-	return gic.client.Images().Delete(gic.GetName(), opts)
+	return gic.configClient.Images().Delete(gic.GetName(), opts)
+}
+
+func (gic *generatorImageConfig) getRouteHostnames() ([]string, error) {
+	externalHostnames := []string{}
+	routes, err := gic.routeLister.List(labels.Everything())
+	if err != nil {
+		return []string{}, err
+	}
+	for _, route := range routes {
+		routeOwner := metav1.GetControllerOf(route)
+
+		// ignore routes that weren't created by the registry operator
+		if routeOwner == nil || routeOwner.UID != gic.owner.UID {
+			continue
+		}
+		for _, ingress := range route.Status.Ingress {
+			hostname := ingress.Host
+			if len(hostname) == 0 {
+				continue
+			}
+			if strings.HasPrefix(hostname, "default-route-"+gic.namespace) {
+				// prepend the default hostname because it has special meaning to the cluster
+				// so we want it first in the list.
+				externalHostnames = append([]string{hostname}, externalHostnames...)
+			} else {
+				externalHostnames = append(externalHostnames, hostname)
```

Comment:
Will they be always in the same order?

Reply:
I assumed so (that a list retrieval of objects always returns those objects in the same order) but maybe @deads2k can confirm.   @deads2k ?

Reply:
> I assumed so (that a list retrieval of objects always returns those objects in the same order) but maybe @deads2k can confirm. @deads2k ?

Yes, we store slices as ordered items and keep their serialization consistent to avoid extra writes.

Reply:
@deads2k not talking about the string slice here though...  asking about the list of Route objects we're getting back from our List API call to the route api.

Reply:
i spoke with david, List results are not guaranteed to return in the same order, so i'll need to add a sort here.